ehrql/query_engines/mssql_dialect.py

```diff
@@ -125,7 +132,42 @@ def get_children(self):
 
 @compiles(SelectStarInto)
 def visit_select_star_into(element, compiler, **kw):
+    source_query = compiler.process(element.selectable.element, asfrom=True, **kw)
+    target = compiler.process(element.table, asfrom=True, **kw)
+
+    # special handling of CTEs
+    if source_query.lower().lstrip().startswith("with"):
+        last_from_ix = source_query.lower().rfind("from")
+        return (
+            f"{source_query[:last_from_ix]} INTO {target} {source_query[last_from_ix:]}"
+        )
+
     return "SELECT * INTO {} FROM {}".format(
-        compiler.process(element.table, asfrom=True, **kw),
+        target,
         compiler.process(element.selectable, asfrom=True, **kw),
     )
+
+
+# implement the "table value constructor trick"
+# https://stackoverflow.com/questions/71022/sql-max-of-multiple-columns/6871572#6871572
+def get_aggregate_subquery(element, aggregate_function):
+    t = check_and_get_function_args_common_type(element)
+    v = values(column("aggregate", t), name="aggregate_values").data(
+        [(c,) for c in element.clauses.clauses]
+    )
+    return sqlalchemy.select(aggregate_function(v.c.aggregate))
+
+
+# N.B SQL Server 2019+ supports GREATEST() and LEAST() so
+# could check version and pass to native function if we
+# ever upgrade
+@compiles(least, "mssql")
+def visit_least(element, compiler, **kw):
+    stmt = get_aggregate_subquery(element, sqlalchemy.func.min).label("least")
+    return compiler.process(stmt, **kw)
```

Comment:
I appreciate that using custom `compiles` constructs is the way the SQLAlchemy documentation will point you, but I actually think it's not what we want to do here because it ends up adding an additional layer of abstraction to the query engine system, which is _already_ an abstraction layer over DBMSs.

So I think the way to handle this would be to have this logic in the MSSQLQueryEngine.

Reply:
I think my latest commit achieves this if you'd be so kind as to take a look